src/d.sea-of-nodes.md

```diff
@@ -0,0 +1,412 @@
+
+  //
+  // `i++`
+  //
+  update:
+   { type: 'UpdateExpression',
+     operator: '++',
+     argument: { type: 'Identifier', name: 'i' },
+     prefix: false },
+
+  //
+  // `arr[i] += 1;`
+  //
+  body:
+   { type: 'ExpressionStatement',
+     expression:
+      { type: 'AssignmentExpression',
+        operator: '+=',
+        left: { type: 'Identifier', name: 'acc' },
+        right:
+         { type: 'MemberExpression',
+           computed: true,
+           object: { type: 'Identifier', name: 'arr' },
+           property: { type: 'Identifier', name: 'i' } } } }
+```
+
+This JSON could also be visualized:
+![AST][7]
+
+This is a tree, so it is very natural to traverse it from the top to the bottom,
+generating the machine code as we visit the AST nodes. The problem with this
+approach is that the information about variables is very sparse, and is spread
+through the different tree nodes.
+
+Again, to safely move the length lookup out of the loop we need to know that the
+array length does not change between the loop's iterations. Humans can do it
+easily just by looking at the source code, but the compiler needs to do quite a
+lot of work to confidently extract those facts directly from the AST.
+
+Like many other compiler problems, this is often solved by lifting the data into
+a more appropriate abstraction layer, i.e. intermediate representation. In this
+particular case that choice of IR is known as a data-flow graph (DFG). Instead
+of talking about syntax-entities (like `for loop`s, `expressions`, ...), we
+should talk about the data itself (read, variables values), and how it changes
+through the program.
+
+## Data-flow Graph
+
+In our particular example, the data we are interested in is the value of
+variable `arr`. We want to be able to easily observe all uses of it to verify
+that there are no out-of-bounds accesses or any other change that would modify
+the length of the array.
+
+This is accomplished by introducing "def-use" (definition and uses) relationship
+between the different data values. Concretely, it means that the value has been
+declared once (_node_), and that it has been used somewhere to create new values
+(_edge_ for every use). Obviously, connecting different values together will
+form a **data-flow graph** like this:
+
+![Data-flow Graph][8]
+
+Note the red `array` box in this vast graph. The solid arrows going out of it
+represent uses of this value. By iterating over those edges, the compiler can
+derive that the value of `array` is used at:
+
+- `loadArrayLength`
+- `checkRange`
+- `load`
+
+Such graphs are constructed in the way that explicitly "clones" the array node,
+if its value was accessed in a destructive manner (i.e. stores, length sizes).
+Whenever we see `array` node and observe its uses - we are always certain that
+its value does not change.
+
+It may sound complicated, but this property of the graph is quite easy to
+achieve. The graph should follow [Single Static Assignment][18] rules.
```

Comment:
It might be nice to have a short version of other rules listed here to give a little more context to the value of SSA.